### ISSUE TITLE
Exports as tasks

### ISSUE THREAD

**Issue by thejameskyle** [gulpjs/gulp#1305]
One of the things I've started doing in all my gulpfiles is to write tasks as plain functions and then write `gulp.task` at the end of the file. This makes abstraction a bit easier and avoids using IDs for composition instead of just functions, making writing large complex gulpfiles much much easier to follow.

```js
function test() { ... }
gulp.task('test', test);
```

This gets a lot easier with gulp 4 using `series` and `parallel`.

```js
function test() { ... }
function buildFiles() { ... }
var build = gulp.series(test, buildFiles);
gulp.task('test', test);
gulp.task('build', build);
```

But at this point writing out `gulp.task` seems a bit unnecessary since the `id` isn't being used in the code and is only there to export a set of tasks.

I'd prefer just to write:

```js
exports.test = test;
exports.build = build;
```

Now this gets even more interesting with ES2015 because you can write:

```js
export function test() { ... }
export function build() { ... }
```

One of the cool benefits here is that you can use static analysis to determine which tasks the `gulpfile` defines and use it for stuff like CLI autocompletion:

```sh
$ gulp [tab]
  - test
  - build
  - dev
```

But the bigger benefit here is that it encourages people to write gulp tasks as plain functions that return streams, making them far more generic and simplifying the gulp API greatly. 

This is what gulpfiles could look like in the future: https://gist.github.com/thejameskyle/2778e4df363612635989

How would you all feel about allowing tasks to be defined as exports?

**Comment by phated** [comment]
@thejameskyle thanks for the feedback.  I've been moving towards what you are suggesting quite heavily, but mainly from a modular gulpfile standpoint.  Basically, make each task function a module and then require them all inside the gulpfile and call gulp.task on the ones that you want exposed to the command line.  I've noticed something in your samples that wouldn't actually work: having a function that returns the function produced by gulp.series wouldn't actually run those functions.  The reason for having people still call gulp.task is because you might want to compose functions as tasks (and it is a bad idea to put those inside other function). e.g.

```js
function test() { ... }
function buildFiles() { ... }
gulp.task(test);
gulp.task('build', gulp.series(test, buildFiles));
```

Note, we use function names as identifiers in gulp 4.0, so there is no need to use strings unless you have an anonymous function (like the composed function from gulp.series/parallel).

**Comment by thejameskyle** [comment]
>  I've noticed something in your samples that wouldn't actually work:

I've fixed the examples

**Comment by thejameskyle** [comment]
Just threw together what a multi-file gulp setup would look like with this https://gist.github.com/thejameskyle/9fd93d3893e2cbd78329

**Comment by phated** [comment]
@thejameskyle there might be a way to support this **just** in the CLI.  I'm thinking we can call `gulp.tree()` and see if the resulting array is empty; if so, call gulp.task on each exported function.  Thoughts?

**Comment by thejameskyle** [comment]
Is there a problem with just adding all the exported functions to the resulting array?

**Comment by phated** [comment]
@thejameskyle yes, lots of problems.  Gulp 4 wasn't created with this use case in mind and it is too much an undertaking to rewrite everything

**Comment by thejameskyle** [comment]
> it is too much an undertaking to rewrite everything

Too bad there isn't an **undertaker** to do this. *shitty rimshot*

That's fine then, I was just wondering.

**Comment by phated** [comment]
@thejameskyle implemented and pinged you in the PR

**Comment by phated** [comment]
Hasn't landed yet ;)

**Comment by thejameskyle** [comment]
Oh sorry, I assumed you would close this when merged

**Comment by phated** [comment]
@thejameskyle yep, I merged to a 1.1 working branch instead of master. Trying to wrap up some features to publish as 1.1

**Comment by phated** [comment]
gulp-cli v1.1.0 has been published with this change.
